File: tests/refs/kth_largest.cpp
// Reference C++ driver for tests/cases/kth_largest.json. Uses a min-heap of size k.
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

long long kthLargest(const std::vector<long long>& a, int k) {
    std::priority_queue<long long, std::vector<long long>, std::greater<long long>> minh;
    for (long long v : a) {
        minh.push(v);
        if ((int)minh.size() > k) minh.pop();
    }
    return minh.top();
}

int main() {
    std::string line, name; std::vector<long long> arr;
    long long k = 0, expected = 0;
    int phase = 0; bool he = false; int p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || phase < 2 || !he) return;
        long long got = kthLargest(arr, (int)k);
        if (got == expected) { std::cout << "PASS kth_largest :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL kth_largest :: " << name << "\n  expected: " << expected << "\n  got: " << got << "\n"; ++f; }
        name.clear(); arr.clear(); phase = 0; he = false;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line); std::string tag; iss >> tag;
        if (tag == "CASE") { iss >> name; phase = 0; }
        else if (tag == "ARR") {
            arr.clear(); std::size_t n; iss >> n;
            for (std::size_t i = 0; i < n; ++i) { long long v; iss >> v; arr.push_back(v); }
            phase = 1;
        } else if (tag == "INT") {
            long long v; iss >> v;
            if (phase == 1) { k = v; phase = 2; } else { expected = v; he = true; }
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/kth_largest.rs
// Reference Rust driver for tests/cases/kth_largest.json. Uses a min-heap of size k.
use std::collections::BinaryHeap;
use std::cmp::Reverse;
use std::io::{self, BufRead};

fn kth_largest(a: &[i64], k: usize) -> i64 {
    let mut h: BinaryHeap<Reverse<i64>> = BinaryHeap::new();
    for &v in a {
        h.push(Reverse(v));
        if h.len() > k { h.pop(); }
    }
    h.peek().map(|r| r.0).unwrap_or(0)
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut arr: Vec<i64> = Vec::new();
    let mut k: i64 = 0; let mut expected: i64 = 0;
    let mut phase = 0i32; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, arr: &mut Vec<i64>, k: i64, expected: i64,
                 phase: &mut i32, he: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *phase >= 2 && *he {
            let got = kth_largest(arr, k as usize);
            if got == expected { println!("PASS kth_largest :: {}", name); *p += 1; }
            else { println!("FAIL kth_largest :: {}\n  expected: {}\n  got: {}", name, expected, got); *f += 1; }
        }
        name.clear(); arr.clear(); *phase = 0; *he = false;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, &mut arr, k, expected, &mut phase, &mut he, &mut p, &mut f); continue; }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { name = parts.next().unwrap_or("").to_string(); phase = 0; }
            "ARR" => {
                arr.clear();
                let n: usize = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                for _ in 0..n { if let Some(t) = parts.next() { if let Ok(v) = t.parse::<i64>() { arr.push(v); } } }
                phase = 1;
            }
            "INT" => {
                let v: i64 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                if phase == 1 { k = v; phase = 2; } else { expected = v; he = true; }
            }
            _ => {}
        }
    }
    flush(&mut name, &mut arr, k, expected, &mut phase, &mut he, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
